openmls/src/credentials/mod.rs

```diff
@@ -187,7 +180,62 @@ pub enum MlsCredentialType {
 #[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
 pub struct Credential {
     credential_type: CredentialType,
-    credential: MlsCredentialType,
+    serialized_credential_content: VLBytes,
+}
+
+impl tls_codec::Size for Credential {
+    fn tls_serialized_len(&self) -> usize {
+        CredentialType::tls_serialized_len(&CredentialType::Basic)
+            + self.serialized_credential_content.as_ref().len()
+    }
+}
+
+impl tls_codec::Serialize for Credential {
+    fn tls_serialize<W: Write>(&self, writer: &mut W) -> Result<usize, Error> {
+        self.credential_type.tls_serialize(writer)?;
+        writer.write_all(self.serialized_credential_content.as_slice())?;
+        Ok(self.tls_serialized_len())
+    }
+}
+
+impl tls_codec::Deserialize for Credential {
+    fn tls_deserialize<R: Read>(bytes: &mut R) -> Result<Self, Error> {
+        // We can not deserialize arbitrary credentials because we don't know
+        // their structure. While we don't care, we still need to parse it
+        // in order to move the reader forward and read the values in the struct
+        // after this credential.
+
+        // The credential type is important, so we read that.
+        let credential_type = CredentialType::tls_deserialize(bytes)?;
+
+        // Now we don't know what we get unfortunately.
+        // We assume that it is a variable-sized vector. This works for the
+        // currently specified credentials and any other credential MUST be
+        // encoded in a vector as well. Otherwise this implementation will
+        // either crash or exhibit unexpected behaviour.
```

Comment:
My understanding is that while parsing can fail with an error, it doesn't have to. Depending on how the credential is encoded, the credential parser could terminate thinking that it has finished and the rest of the bytes that are actually part of the credential are interpreted as something else by the parser of the next field in the message. The behaviour of OpenMLS then depends on what all that looks like. It shouldn't be a security issue (because we still have signatures), but it could lead to errors in places that have nothing to do with the credential.

We might want to be more precise w.r.t. what we mean by crashes. I think the worst case is that OpenMLS will throw an error that is otherwise unrelated to the credential.

Reply:
Yep, what Konrad said. I rephrased slightly.
There is undefined behaviour here that we can not avoid. This is the trade-off for allowing other credentials and not refactoring everything for using generic credentials throughout (and figuring out a way to handle different credentials in a group).